### PR TITLE
[#13] [Integrate] As a user, I can see the list of surveys 

### PR DESCRIPTION
**Pull request by Tuubz**
Closes https://github.com/nimblehq/ic-flutter-taher-toby/issues/13

## What happened 👀

- Add a view model with state handling
- Show surveys with data from API (if succeeded)
- Show snack bar with an error message from API (if failed)
- Add unit tests for the view model 

## Insight 📝

First of all, two dependencies are critical to our MVVM architecture:
- Add [riverpod](https://pub.dev/packages/riverpod) to manage states
   → Usage of `Stream`, `Providers` & `State`
- Add [rx-dart](https://pub.dev/packages/rxdart) to manage asynchronous handling 
   → Usage of `BehaviorSubject` & `PublishSubject`

Second, we align the circle indicator with the page viewer by using [ValueNotifier & ValueListenableBuilder](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html)

I would highly suggest studying these resources of Riverpod:
- ["What are providers?"](https://riverpod.dev/docs/concepts/providers) 
- ["How to read providers?"](https://riverpod.dev/docs/concepts/reading)

**Note**: I temporarily removed `_doRefreshToken` because it's running into an infinite loop when unauthorized

We can look into it later, but for now, we must add _ignore_ comments for the linter to pass ✔️ 

https://user-images.githubusercontent.com/18277915/229686184-1127737d-bf36-400c-bb1f-82e978e85f37.mov

## Proof Of Work 📹

**SUCCESS CASE:**

https://user-images.githubusercontent.com/18277915/229402813-b19d49c3-94b1-4c4e-8812-f66a6adc39b5.mov

**FAILED CASE:**

https://user-images.githubusercontent.com/18277915/229687053-87df203f-a44c-4388-8a7d-f2d5e37d7c9a.mov

### CODE REVIEW

File: lib/model/survey_model.dart [hunk 1]
```diff
@@ -4,13 +4,23 @@ import 'package:flutter_survey/api/response/survey_response.dart';
...
   final String title;
+  final String description;
+  final String coverImageUrl;
 
-  const SurveyModel({required this.id, required this.title});
+  const SurveyModel(
+      {required this.id,
+      required this.title,
+      required this.description,
+      required this.coverImageUrl});
```
Comment (Shayokh144) [inline lib/model/survey_model.dart:14]:
```suggestion
      required this.coverImageUrl}
      );
```

File: lib/model/survey_model.dart [hunk 2]
```diff
@@ -4,13 +4,23 @@ import 'package:flutter_survey/api/response/survey_response.dart';
...
-  List<Object?> get props => [id, title];
+  List<Object?> get props => [id, title, description, coverImageUrl];
 
   factory SurveyModel.fromResponse(SurveyResponse response) {
-    return SurveyModel(id: response.id ?? "", title: response.title ?? "");
+    return SurveyModel(
+        id: response.id ?? "",
+        title: response.title ?? "",
+        description: response.description ?? "",
+        coverImageUrl: response.getHdCoverImageUrl());
```
Comment (Shayokh144) [inline lib/model/survey_model.dart:24]:
```suggestion
        coverImageUrl: response.getHdCoverImageUrl()
        );
```

File: lib/api/response/survey_response.dart [hunk 3]
```diff
@@ -6,13 +7,25 @@ part 'survey_response.g.dart';
 class SurveyResponse {
   String? id;
   String? title;
+  String? description;
+  String? coverImageUrl;
```
Comment (luongvo) [inline lib/api/response/survey_response.dart:11]: 👍

File: lib/di/interceptor/app_interceptor.dart [hunk 4]
```diff
@@ -1,3 +1,4 @@
+// ignore: unused_import
```
Comment (luongvo) [inline lib/di/interceptor/app_interceptor.dart:1]: What does this mean? Do we need to let the linter skip checking import here?

File: lib/di/interceptor/app_interceptor.dart [hunk 5]
```diff
@@ -26,27 +27,25 @@ class AppInterceptor extends Interceptor {
...
-    }
+    handler.next(err);
+    // TODO: Integrate refresh-token https://github.com/nimblehq/ic-flutter-taher-toby/issues/20
+    // final statusCode = err.response?.statusCode;
+    // if ((statusCode == HttpStatus.forbidden ||
+    //         statusCode == HttpStatus.unauthorized) &&
+    //     _requireAuthentication) {
+    //   _doRefreshToken(err, handler);
+    // } else {
+    // }
```
Comment (luongvo) [inline lib/di/interceptor/app_interceptor.dart:38]: Why do we need to disable this, please?

File: lib/ui/home/home_screen.dart [hunk 6]
```diff
@@ -1,19 +1,83 @@
...
+    if (errorMessage.isNotEmpty) showSnackBar(context, errorMessage);
+    return Scaffold(
+      body: Stack(
+        children: [
+          if (surveys.isNotEmpty) ...[
+            HomeSurveyPageViewer(
+              surveys: surveys,
+              currentPage: _currentPage,
+            ),
+            const HomeHeader(),
```
Comment (luongvo) [inline lib/ui/home/home_screen.dart:67]: The header should be shown even if the app can or can't load the survey list. The header depends on the `/me` API call.

File: lib/ui/home/widget/home_survey_page.dart [hunk 7]
```diff
@@ -1,26 +1,27 @@
...
+    required this.survey,
     required this.onNextButtonPressed,
   }) : super(key: key);
 
   @override
   Widget build(BuildContext context) {
     return Stack(
       children: [
-        DimmedBackground(background: Assets.images.dummyBackground.path),
+        DimmedBackground(background: survey.coverImageUrl),
```
Comment (luongvo) [inline lib/ui/home/widget/home_survey_page.dart:24]: Should we use hd image?

**Review by luongvo — COMMENTED** [review]
The rest lgtm

File: lib/di/interceptor/app_interceptor.dart (hunk 5 above)
Comment (Tuubz) [reply lib/di/interceptor/app_interceptor.dart:38]: @luongvo Seems like it's going into an infinite loop of requests when the token is unauthorized

So I remove the refresh token function for now, so we can look into it deeper later

https://user-images.githubusercontent.com/18277915/229685785-bd872ea1-f1a2-4126-8f2c-592f16cb0ab1.mov


I'll update the insights 🙏

File: lib/di/interceptor/app_interceptor.dart (hunk 4 above)
Comment (Tuubz) [reply lib/di/interceptor/app_interceptor.dart:1]: Yes, correct. 

It's related to the comment below 🙏 

I've updated the insights 🚀

File: lib/ui/home/home_screen.dart (hunk 6 above)
Comment (Tuubz) [reply lib/ui/home/home_screen.dart:67]: Thank you,  I'll move `HomeHeader` out of the `if (surveys.isNotEmpty) ...[` condition 🙏

File: lib/ui/home/widget/home_survey_page.dart (hunk 7 above)
Comment (Tuubz) [reply lib/ui/home/widget/home_survey_page.dart:24]: This is already the case, it's being mapped in `survey_model.dart` - `getHdCoverImageUrl()`

**Comment by Tuubz** [comment]
@Shayokh144 @luongvo Updated code and PoW 🚀

**Review by luongvo — APPROVED** [review]

**Review by Shayokh144 — APPROVED** [review]

**Pull request merged**